hugr-core/src/std_extensions/collections/array/array_clone.rs

```diff
@@ -0,0 +1,224 @@
+    fn name(&self) -> OpName {
+        ARRAY_CLONE_OP_ID
+    }
+}
+
+impl<AK: ArrayKind> FromStr for GenericArrayCloneDef<AK> {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s == ARRAY_CLONE_OP_ID {
+            Ok(GenericArrayCloneDef::new())
+        } else {
+            Err(())
+        }
+    }
+}
+
+impl<AK: ArrayKind> GenericArrayCloneDef<AK> {
+    /// To avoid recursion when defining the extension, take the type definition as an argument.
+    fn signature_from_def(&self, array_def: &TypeDef) -> SignatureFunc {
+        let params = vec![TypeParam::max_nat(), TypeBound::Copyable.into()];
+        let size = TypeArg::new_var_use(0, TypeParam::max_nat());
+        let element_ty = Type::new_var_use(1, TypeBound::Copyable);
+        let array_ty = AK::instantiate_ty(array_def, size, element_ty)
+            .expect("Array type instantiation failed");
+        PolyFuncTypeRV::new(
+            params,
+            FuncValueType::new(array_ty.clone(), vec![array_ty; 2]),
+        )
+        .into()
+    }
+}
+
+impl<AK: ArrayKind> MakeOpDef for GenericArrayCloneDef<AK> {
+    fn from_def(op_def: &OpDef) -> Result<Self, OpLoadError>
+    where
+        Self: Sized,
+    {
+        crate::extension::simple_op::try_from_name(op_def.name(), op_def.extension_id())
+    }
+
+    fn init_signature(&self, _extension_ref: &Weak<Extension>) -> SignatureFunc {
+        self.signature_from_def(AK::type_def())
+    }
+
+    fn extension_ref(&self) -> Weak<Extension> {
+        Arc::downgrade(AK::extension())
+    }
+
+    fn extension(&self) -> ExtensionId {
+        AK::EXTENSION_ID
+    }
+
+    fn description(&self) -> String {
+        "Clones an array with copyable elements".into()
+    }
+
+    /// Add an operation implemented as a [MakeOpDef], which can provide the data
+    /// required to define an [OpDef], to an extension.
+    //
+    // This method is re-defined here since we need to pass the array type def while
+    // computing the signature, to avoid recursive loops initializing the extension.
+    fn add_to_extension(
+        &self,
+        extension: &mut Extension,
+        extension_ref: &Weak<Extension>,
+    ) -> Result<(), crate::extension::ExtensionBuildError> {
+        let sig = self.signature_from_def(extension.get_type(&AK::TYPE_NAME).unwrap());
+        let def = extension.add_op(self.name(), self.description(), sig, extension_ref)?;
+        self.post_opdef(def);
+        Ok(())
+    }
+}
+
+/// Definition of the array clone op. Generic over the concrete array implementation.
+#[derive(Clone, Debug, PartialEq)]
+pub struct GenericArrayClone<AK: ArrayKind> {
+    /// The element type of the array.
+    pub elem_ty: Type,
+    /// Size of the array.
+    pub size: u64,
+    _kind: PhantomData<AK>,
+}
+
+impl<AK: ArrayKind> GenericArrayClone<AK> {
+    /// Creates a new array clone op.
+    pub fn new(elem_ty: Type, size: u64) -> Option<Self> {
```

Comment:
should this return `Result` rather than option?

at the very least docstring should explain the return type